### ISSUE TITLE
webpack@4, node@13, Mac crashes when compiler.watch().close()

### ISSUE THREAD

**Issue by AviVahl** [webpack/webpack#10037]
# Bug report
When closing a watching compiler programmatically, the node process crashes.

**What is the current behavior?**
Getting the following crash in several repositories on our Mac CIs:
```
Assertion failed: (0), function uv_close, file ../deps/uv/src/unix/core.c, line 187.
error Command failed with signal "SIGABRT".
```

**If the current behavior is a bug, please provide the steps to reproduce.**
- clone repo: https://github.com/AviVahl/lodash-es-webpack4
- `yarn`
- paste the following to `a.js`:
```js
const webpack = require('webpack');
const compiler = webpack({ mode: 'development' });

function createWatcher() {
    const watcher = compiler.watch({}, () => {
        console.log('watching');
        setTimeout(() => {
            watcher.close(() => {
                console.log('closed watcher');
                createWatcher();
            });
        }, 1000);
    });
}

createWatcher();
```
- `node a.js`

**What is the expected behavior?**
(for the above code) watch -> wait for 1 sec -> close watcher -> rewatch and repeat process.

**Other relevant information:**
webpack version:  4.41.2
Node.js version: 13.2.0
Operating System: Mac 10.12.6

@sokra maybe this is because `watchpack` is using old `chokidar` with outdated `fsevents`? :/

**Comment by MarcGaumont** [comment]
Have the exact same problem...

**Comment by AviVahl** [comment]
To be honest, I wanted to re-check whether this is still an issue with fsevents@1.2.11 (which added Node 12 compatibility)... I use a Linux machine, so I need to get my hands on a Mac.

**Comment by AviVahl** [comment]
I've re-checked the above instructions... still crashes on Mac and Node 13.

**Comment by evilebottnawi** [comment]
What is version node? `13.0`, `13.1`? Looks like bug on node side

**Comment by PandaWhisperer** [comment]
Checking in from the Nuxt.js issue that was just linked. Node.js 13.5.0 with macOS 10.15.2 here.

**Comment by KrimoPixelsmith** [comment]
Same issue here node 13.2, npm 6.13.4, macOS 10.15.2

**Comment by evilebottnawi** [comment]
Can you try to update node to latest version?

**Comment by AviVahl** [comment]
just re-tested on Node 13.6.0 with latest `webpack`. re-produced the issue consistently.

Btw, I'm not certain it's a Node issue. Mac is the only platform where `watchpack` doesn't use native node APIs and instead uses `fsevents` (via `chokidar`). The issue could be there...

**Comment by KrimoPixelsmith** [comment]
@evilebottnawi just like @AviVahl said, it's the exact same; from what I could tell looking online, it looks like it's a chodikar/fsevents issue; I first noticed it in using https://github.com/JeffreyWay/laravel-mix, which uses chodikar version 2.0.x and webpack 4 still;

Looks like v3 is out...

**Comment by PandaWhisperer** [comment]
Can confirm. Updated to node 13.6.0 and the error still occurs.

**Comment by mahnunchik** [comment]
The same error. Node v13.6.0

**Comment by vjpr** [comment]
I am testing with v13.8 now.

**Comment by 7iomka** [comment]
```
Assertion failed: (0), function uv_close, file ../deps/uv/src/unix/core.c, line 187.
Abort trap: 6
$ node -v
v13.6.0
```
:(

**Comment by DevJonTaylor** [comment]
node v13.8.0
macOS 10.15.2 (19C57)
When hot-loading an update it does not crash with this error.
I tested with scss-node | html | and javascript using *.js, *.vue, and *.scss files.

However when closing the node I receive the error again.

```
 DONE  Compiled successfully in 1153ms                                                                                                      9:15:29 AM


  App running at:
  - Local:   http://localhost:8080/ 
  - Network: http://192.168.1.11:8080/

^CAssertion failed: (0), function uv_close, file ../deps/uv/src/unix/core.c, line 187.
zsh: abort      npm run serve
jontaylor@Jonathans-MacBook-Pro portfolio % 
```

**Comment by sokra** [comment]
For webpack 4 we can't upgrade to chokidar 3, as this would be a breaking change. This would need to be fixed in a chokidar for webpack 4 to benefit from it. I also see these errors randomly occuring in the CI build.

For webpack 5 we migrated away from chokidar and native dependencies. I no longer see these errors in the webpack CI builds.

**Comment by kelly-tock** [comment]
i'm getting this as well on node 13.8 with karma and karma webpack. was fine on node v11.12

**Comment by Nikodermus** [comment]
Failing as well

```
node v13.8.0
macos mojave 10.14 (18A391)
```

**Comment by gansdot** [comment]
Am also getting the below error. npm server crashes after few mins connecting to mongo db. any idea why this is ?

Assertion failed: (0), function uv_close, file ../deps/uv/src/unix/core.c, line 187.
zsh: abort      npm run dev

**Comment by evilebottnawi** [comment]
Anyone experiencing a problem please read - https://github.com/fsevents/fsevents/issues/314

Briefly - not fixable, not sure if even I send a fix for it, it will be merged.

Good news - webpack@5 doesn't use `fsevents`, so no problems

**Comment by AviVahl** [comment]
That's a shame, because this started failing with the newly released Node 12.16.2 as well. :(

**Comment by evilebottnawi** [comment]
@AviVahl You can open a new issue, maybe when the number of issues becomes more, they will change their minds

**Comment by aminebenkeroum** [comment]
Same experience : Node v13.8.0, Mac Catalina OS

**Comment by toasterpic** [comment]
Same here, node 13.13.0, MacOS Mojave 10.14.6 (18G87)

**Comment by slgarkea** [comment]
I fixed it by forcing `chokidar` in his last version :
```
  "resolutions": {
    "chokidar": "2.1.8"
  }
```
No anormal behaviour seen so far.

**Comment by edi** [comment]
Same here (using Expo bare-minimum).
Node v13.6.0
Mac Catalina OS

**Comment by slgarkea** [comment]
For those using React Native, maybe this fix could help : https://github.com/facebook/react-native/commit/0d1fb458ab8027dcfac5f2fa11e8c16d6853c59c ?

I don't have any issue with iOS compilation but Android's one keep failing.

**Comment by KrimoPixelsmith** [comment]
> Edit: nope, this not work. This is weird, some build finishes successfully randomly.
> 
> I fixed it by forcing `chokidar` in his last version :
> 
> ```
>   "resolutions": {
>     "chokidar": "3.3.1"
>   }
> ```
> 
> No anormal behaviour seen so far.

Same exact issue as you Sylvain (et bonjour d'un Normand au passage!), forcing chokidar to the latest version seemed to fix it, but only sometimes.

I've tested this with Vue CLI 3 and Laravel Mix 5; on both occasion the problem seems to occur when compiling scss.

**Comment by evilebottnawi** [comment]
@KrimoPixelsmith what is a error message?

**Comment by KrimoPixelsmith** [comment]
@evilebottnawi same errror as before,
`Assertion failed: (0), function uv_close, file ../deps/uv/src/unix/core.c, line 187.`

**Comment by evilebottnawi** [comment]
@KrimoPixelsmith hm, we need fix api for close, because chokidar uses async `close`, before it was sync

**Comment by slgarkea** [comment]
I update node to the last version (14.1.0 - prior was 13.1.0) and the issue is gone.
But i don't know why.

**Comment by allaniftrue** [comment]
I'm using `node v12.16.3 (npm v6.14.4)` and i still get the same error

`Assertion failed: (0), function uv_close, file ../deps/uv/src/unix/core.c`

**Comment by AviVahl** [comment]
Issue got replaced with another:
```
> fsevents@1.2.13 install /Users/runner/runners/2.169.1/work/mocha-pup/mocha-pup/node_modules/watchpack/node_modules/fsevents
> node install.js

internal/modules/cjs/loader.js:638
    throw err;
    ^

Error: Cannot find module 'nan'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:636:15)
    at Function.Module._load (internal/modules/cjs/loader.js:562:25)
    at Module.require (internal/modules/cjs/loader.js:692:17)
    at require (internal/modules/cjs/helpers.js:25:18)
    at [eval]:1:1
    at Script.runInThisContext (vm.js:122:20)
    at Object.runInThisContext (vm.js:329:38)
    at Object.<anonymous> ([eval]-wrapper:6:22)
    at Module._compile (internal/modules/cjs/loader.js:778:30)
    at evalScript (internal/bootstrap/node.js:590:27)
gyp: Call to 'node -e "require('nan')"' returned exit status 1 while in binding.gyp. while trying to load binding.gyp
gyp ERR! configure error 
gyp ERR! stack Error: `gyp` failed with exit code: 1
gyp ERR! stack     at ChildProcess.onCpExit (/Users/runner/hostedtoolcache/node/10.20.1/x64/lib/node_modules/npm/node_modules/node-gyp/lib/configure.js:351:16)
gyp ERR! stack     at ChildProcess.emit (events.js:198:13)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:248:12)
gyp ERR! System Darwin 19.4.0
gyp ERR! command "/Users/runner/hostedtoolcache/node/10.20.1/x64/bin/node" "/Users/runner/hostedtoolcache/node/10.20.1/x64/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /Users/runner/runners/2.169.1/work/mocha-pup/mocha-pup/node_modules/watchpack/node_modules/fsevents
gyp ERR! node -v v10.20.1
gyp ERR! node-gyp -v v5.1.0
gyp ERR! not ok 
```
Happens for [mocha-pup](https://github.com/wixplosives/mocha-pup) for all node 10/12/14 on Mac.

The new issue appears to already be documented here: https://github.com/fsevents/fsevents/issues/325

Now's the twist. when I tried to reproduce in a separate project using Node v12.16.3 (npm v6.14.4), I've encountered what seems to be a bug in how npm handles optional dependencies.

If I:
```sh
mkdir testing
cd testing
npm init -y
npm i webpack
```
And then look in the `package-lock.json` file, I can see that `fsevents` has been resolved to v1.2.13, and that it requires `nan`:
```json
    "fsevents": {
      "version": "1.2.13",
      "resolved": "https://registry.npmjs.org/fsevents/-/fsevents-1.2.13.tgz",
      "integrity": "sha512-oWb1Z6mkHIskLzEJ/XWX0srkpkTQ7vaopMQkyaEIoq0fmtFVxOthb8cCxeT+p3ynTdkk/RZwbgG4brR5BeWECw==",
      "optional": true,
      "requires": {
        "bindings": "^1.5.0",
        "nan": "^2.12.1"
      }
```
`nan` is also found is the lock file.

But if I run the same set of commands, just adding chokidar to the install command:
```sh
mkdir testing
cd testing
npm init -y
npm i webpack chokidar
```

Then the `package-lock.json` contains:
```json
        "fsevents": {
          "version": "1.2.13",
          "resolved": "https://registry.npmjs.org/fsevents/-/fsevents-1.2.13.tgz",
          "integrity": "sha512-oWb1Z6mkHIskLzEJ/XWX0srkpkTQ7vaopMQkyaEIoq0fmtFVxOthb8cCxeT+p3ynTdkk/RZwbgG4brR5BeWECw==",
          "optional": true
        },
```
Without the `"requires"`, and without `nan` appearing anywhere in the lock file.

It seems the existence of another `fsevents` version in the tree affects how the dependencies of `fsevents@1` are resolved by npm, leading to the missing `nan` install errors.

The above behavior reproduces consistently (try it yourself) and leads me to believe that depending on which dependencies are in your project, you might experience one of two behaviors:
- Error during `postinstall` of `fsevents` just on Mac (on all node versions), which doesn't fail `npm install`, being an `optionalDependency`. There would be no runtime crashes, as `fsevents` isn't used.
- No errors during `postinstall` anywhere, but runtime crash when programmatically closing the watch compiler on Node@14/Mac and _maybe_ on latest Node@12 as well.

This also accounts for why some people no longer see the crash, and others do (in comments above).

**Comment by paulmillr** [comment]
You should ask maintainers to upgrade to fsevents v2.

**Comment by AviVahl** [comment]
@paulmillr already answered is in this thread: https://github.com/webpack/webpack/issues/10037#issuecomment-583964274

`webpack@4` needs to support node@6, which newer `chokidar/fsevents` does not. I can't really argue with that need. I'm just stuck until `webpack` releases v5 _and_ the loader/plugin ecosystem adapts to the new version.

Well, TBH, not really stuck... we've added workarounds in several of our code-bases to force no watching when in "single-run" mode... :/ This can get a bit hairy, because sometimes it's abstractions like `webpack-dev-middleware` which call the watcher. We ended up monkey patching the webpack's compiler so every call to `.watch()` will call `.run()` instead:
https://github.com/wixplosives/engine/blob/v13.0.2/packages/scripts/src/application.ts#L186
https://github.com/wixplosives/mocha-pup/blob/v3.0.0/src/run-tests.ts#L43

These are just two examples from open source projects.

I hate these workarounds, but we are already using Node 12 in production and started supporting 14. We cannot be held back by such issues.

**Comment by paulmillr** [comment]
@AviVahl not really. There is an effort to upgrade to chokidar 3, maintainers merged pull request/s already. Ask them to release new versions asap.

**Comment by AviVahl** [comment]
Alright, so the latest releases of `watchpack@1` should fix the crash. How, might you ask? It's installing **two** versions of chokidar with a try/catch to each (newer version gets higher priority). I'll save what I really think about this approach, and just say the "postinstall" errors we get from chokidar@2 are *far* from ideal. :(

**Comment by sokra** [comment]
@AviVahl I also dislike that. Technically chokidar2 should only be installed when node<8, but npm seem to ignore that. Works fine in yarn.
We will get rid of that in the next major version, but for semver reasons we need to support node 6 in watchpack 1 and webpack 4. (I know that it is out of lifetime, but that doesn't allow us to break semver here)

**Comment by paulmillr** [comment]
@sokra see [this comment](https://github.com/webpack/watchpack/issues/130#issuecomment-624298549) on "removing chokidar in next major version". The file watching task is not as simple as it looks.

**Comment by AviVahl** [comment]
@sokra both npm@6.14.4 (comes with latest node 12/14) and yarn@1.22.4 (current "latest" for yarn) warn about and install duplicate `chokidar` and `fsevents` versions. I initialized an empty project and just `npm i watchpack` / `yarn add watchpack`. So the statement about any of them being ignored may be incorrect.

Alternative possible solution:
Re-release latest webpack@4 as 5, breaking *just* min Node version. Re-direct current 5 to 6. Supporting it in loaders/plugins would be **easy**, as there would be no other breaking changes. This is probably the "proper" solution, and will allow webpack to both keep Node 6 support and have clean Node 12/14 support.

All this juggling of dependency versions is starting to get out of hand. You end up using older versions of packages just to keep supporting an *unsupported* Node version.

Meanwhile, consumer projects using LTS versions of Node end up having [duplicate dependencies](https://github.com/webpack/webpack/issues/10037#issuecomment-629684336), weird [post-install errors](https://github.com/webpack/webpack/issues/10037#issuecomment-625515136), and failing CI builds due to runtime errors (original issue).

I mean, we've been dealing with this for quite a while now. Consumer projects started having the original issue over 6 months ago.

Having said the above, I wanna add that I appreciate your hard work, and will respect whatever decision you end up taking here.

**Comment by briandipalma** [comment]
We're not going to run out of numbers any time soon, releasing Webpack 5 just to drop node 6 support seems reasonable.

**Comment by AviVahl** [comment]
Alright, I am closing this issue, as the reported crash should now be fixed (make sure your project has watchpack resolved to v1.7.2 or above).

There's a [separate new issue](https://github.com/webpack/webpack/issues/10037#issuecomment-625515136) (postinstallation errors) which is expected to resolve itself with `webpack@5`.

**Comment by JulienKode** [comment]
I ad the same issue with

Node: v12.16.3
OSX: 10.15.5
